### DISCUSSION TITLE
jsPsychSurveyHtmlForm (save inputs)

### DISCUSSION THREAD

**Discussion by Lena3558** [jspsych/jsPsych#3202] [Q&A]
Hello,
I'm new to jspsych and I have run into a problem while coding a simple online questionnaire. I would be very thankful for any help.

    var jsPsych = initJsPsych({
      on_finish: function(data) {
	    var falseline = [];

        for (var i = 1; i <= 13; i++) {
		  var leftInput = jsPsych.data.get().select('response').values[0]['L' + i].length;
          var rightInput = jsPsych.data.get().select('response').values[0]['R' + i].length;
		  if (leftInput + rightInput > 2) {
		    falseline.push(i);
          }
        }

        if (falseline.length > 0) {
		  var Message = "Items: " + falseline.join(', ');
          alert(Message);
		  // Problem: refresh page without removing the input
        } else {
          jsPsych.data.displayData();
        }
      }
    });

	var timeline = [];
	
	var ehi = {
      type: jsPsychSurveyHtmlForm,
      preamble: '<p> Instructions</p>',
	  html: '\
	    <pre><p>1. Schreiben Links <input name="L1" type="text" maxlength="2"/>   Rechts <input name="R1" type="text" maxlength="2"/></pre></p>\
	    <pre><p>2. Zeichnen Links <input name="L2" type="text" maxlength="2"/>   Rechts <input name="R2" type="text" maxlength="2"/></pre></p>\
	  /...
	  ',
	  button_label: 'Weiter',
	};
	timeline.push(ehi);

This was my approach. I want to present all the items and check if there are more than 2 inputs per line (e.g. L1 + R1). If this is the case the alert message appears, but when I refresh the page, all input is removed and the participants have to redo everything. Is there a way to save the inputs? Or is there an alternative way to solve this problem?

Thank you very much!
Lena

**Comment by Shaobin-Jiang** [comment]
I do not think there is any easy way to do this. If you are running the experiment online and can be certain that the experiment is going to be run only **once** on each participant's device, then you can probably use [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage). You can also take a look at the tutorial [here](https://javascript.info/localstorage). I am not saying that is the prerequisite of `localStorage`, but if you are running the experiment repeatedly on one device, then you will have to distinguish between the fresh start of an experiment and the reloading of a previous session.

Another way is using jsPsych's child timeline and [`loop_function`](https://www.jspsych.org/7.3/overview/timeline/#looping-timelines). You can put your trials in a child timeline, run it once, check whether invalid inputs exist, and if they do, loop the timeline. In the next run, you can fill the input elements with the values from the previous run. This can lead to more data than is probably desired, but it should work fine under all occasions.

**Reply by jodeleeuw** [reply]
I would second the loop_function solution. I think that's the simplest without modifying the plugin itself.

**Reply by Lena3558** [reply]
Thanks a lot for your help!
